### PR TITLE
Feature/UI metrics view

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Adds internal forwarding of metrics.
Adds standalone and cluster views for metrics (coordinator and dbserver specific in cluster case)
Standalone: Dashboard -> Metrics
Cluster: Nodes -> Select a node -> Metrics 

- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)

#### Backports:

- [x] No backports required

#### Related Information

https://arangodb.atlassian.net/browse/GORDO-1031

### Testing & Verification

*(Please pick either of the following options)*

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### CODE REVIEW

**Comment by hkernbach** [comment]
@jsteemann currently it seems that API does not provide server specific metrics. Need to check this after the weekend.

File: arangod/CMakeLists.txt [hunk 1]
```diff
@@ -967,6 +967,7 @@ target_link_libraries(arango_iresearch arango_indexes)
 target_link_libraries(arango_iresearch arango_cluster_engine)
 
 target_link_libraries(arango_metrics arango)
+target_link_libraries(arango_metrics arango::validation)
```
Comment (jsteemann) [inline arangod/CMakeLists.txt:970]: Wondering what this change has to do with the metrics UI changes... :thinking:

File: arangod/RestServer/Metrics.cpp [hunk 2]
```diff
@@ -102,8 +102,8 @@ void Counter::store(uint64_t const& n) {
 
 void Counter::toPrometheus(std::string& result) const {
   _b.push();
-  result += "\n#TYPE " + name() + " counter\n";
-  result += "#HELP " + name() + " " + help() + "\n";
+  result += "\n# HELP " + name() + " " + help() + "\n";
+  result += "# TYPE " + name() + " counter\n";
```
Comment (jsteemann) [inline arangod/RestServer/Metrics.cpp:106]: Ok... why do we need to swap the order of `TYPE` and `HELP` here, and why do we need the extra space?
According to https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#grouping-and-sorting these changes are valid, but not necessary. 
To keep our API output as close to previous ArangoDB versions as possible, I would favor reverting this change.

File: arangod/V8Server/v8-statistics.cpp [hunk 3]
```diff
@@ -218,6 +218,21 @@ static void JS_EnabledStatisticsAllDatabases(v8::FunctionCallbackInfo<v8::Value>
...
+
+static void JS_EnabledMetrics(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate)
+    v8::HandleScope scope(isolate);
+
+    TRI_GET_GLOBALS();
+    bool enabled = v8g->_server.getFeature<MetricsFeature>().exportAPI();
+    v8::Handle<v8::Value> result = v8::Boolean::New(isolate, enabled);
+    TRI_V8_RETURN(result);
+  TRI_V8_TRY_CATCH_END
```
Comment (jsteemann) [inline arangod/V8Server/v8-statistics.cpp:233]:
```suggestion
  TRI_V8_TRY_CATCH_BEGIN(isolate)
  v8::HandleScope scope(isolate);

  TRI_GET_GLOBALS();
  bool enabled = v8g->_server.getFeature<MetricsFeature>().exportAPI();
  v8::Handle<v8::Value> result = v8::Boolean::New(isolate, enabled);
  TRI_V8_RETURN(result);
  TRI_V8_TRY_CATCH_END
```

File: js/apps/system/_admin/aardvark/APP/frontend/js/arango/arango.js [hunk 4]
```diff
@@ -1066,9 +1066,10 @@
       var dlType;
       if (type === 'csv') {
         dlType = 'text/csv; charset=utf-8';
-      }
-      if (type === 'json') {
+      } else if (type === 'json') {
         dlType = 'application/json; charset=utf-8';
+      } else if (type === 'text') {
+        dlType = 'text/html; charset=utf8';
```
Comment (jsteemann) [inline js/apps/system/_admin/aardvark/APP/frontend/js/arango/arango.js:1072]: If this is for our metrics out, I think the correct Content-Type would be `text/plain`, not `text/html`.

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/documentsView.js [hunk 5]
```diff
@@ -1090,7 +1090,12 @@
       var self = event.currentTarget;
 
       var url;
-      var doc = $(self).attr('id').substr(4);
+      var doc;
+      if ($(self).attr('id')) {
+        doc = $(self).attr('id').substr(4);
+      } else {
+        return;
+      }
```
Comment (jsteemann) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/documentsView.js:1098]: The following would be easier to read:
```suggestion
      if (!$(self).attr('id')) {
        return;
      }
      var doc = $(self).attr('id').substr(4);
```

File: js/apps/system/_admin/aardvark/APP/frontend/js/arango/arango.js (hunk 4 above)
Comment (hkernbach) [reply js/apps/system/_admin/aardvark/APP/frontend/js/arango/arango.js]: Done, ty.

File: arangod/RestServer/Metrics.cpp (hunk 2 above)
Comment (hkernbach) [reply arangod/RestServer/Metrics.cpp:106]: Well, yes true. The definition is a bit clumsy. I just wanted to stick to the examples given there. 
Also: Reading name + definition first makes more sense from a readers perspective than reading the type first.
Additionally, libraries I've found do want that extra space for parsing. If someone is not using prom UI, and wanted to use those libraries for parsing as well, they will run into the same problems I ran into. 

So I vote for keeping those changes ( at least the extra space - but IMHO also the order of attributes)

File: arangod/CMakeLists.txt (hunk 1 above)
Comment (hkernbach) [reply arangod/CMakeLists.txt:970]: Without that addition, the target arango_metrics is not able to compile.

**Pull request merged**